File: dsh-manhua/src/story.ts
import fs from "node:fs";
import path from "node:path";
import { STORIES_DIR, ensureDir, loadConfig } from "./config.js";
import { REPO_ROOT } from "./paths.js";
import { ok, warn } from "./zh-log.js";

export type RefImages = Record<string, string>;

export type StoryCharacter = {
  id: string;
  name: string;
  identity_lock: string;
  role?: string;
  sheet_prompt?: string;
  ref_images?: RefImages;
  /** false 时禁止下游成片；缺省视为已批准 */
  approved?: boolean;
};

export type StoryEnvironment = {
  id: string;
  name: string;
  scene_card: string;
  time_of_day?: string;
  weather?: string;
  ref_images?: RefImages;
};

export type StoryProp = {
  id: string;
  name: string;
  clue_lock: string;
  ref_images?: RefImages;
};

export type StoryShot = {
  shot_id: string;
  environment_id: string;
  character_ids: string[];
  action: string;
  episode_id?: string;
  scene_id?: string;
  beat_id?: string;
  prop_ids?: string[];
  duration?: number;
  still_prompt?: string;
  still_url?: string;
  video_prompt?: string;
  ref_images?: string[];
  ref_audios?: string[];
  camera?: Record<string, string>;
  emotion?: string;
  dialogue?: string | null;
  plan_path?: "video_ref" | "bridge" | "grid" | "lipsync" | string;
  plan_notes?: string;
  bridge_from?: string | null;
  first_frame?: string;
  last_frame?: string;
  needs_lipsync?: boolean;
  still_approved?: boolean;
  video_approved?: boolean;
  segment_break?: boolean;
  hard_cut?: boolean;
  grid_cell?: number;
  workflow?: string;
  seed?: number | null;
  state?: Record<string, unknown>;
  [k: string]: unknown;
};

export type StoryPack = {
  story_id: string;
  chapter_id: string;
  project_id: string;
  title?: string;
  style_lock?: string;
  resolution?: string;
  video_workflow?: string;
  steps?: Array<"audio" | "video" | "bridge" | "lipsync" | string>;
  /** 章节管理新建的草稿章：允许空分镜，待后续流程填充 */
  draft?: boolean;
  script: Record<string, unknown>;
  characters: StoryCharacter[];
  environments: StoryEnvironment[];
  props?: StoryProp[];
  shots: StoryShot[];
  [k: string]: unknown;
};

export class StoryError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "StoryError";
  }
}

export function storyPath(storyId: string, chapterId?: string): string {
  if (chapterId?.trim()) {
    const cid = chapterId.trim();
    const chapterFile = chapterFilePath(storyId, cid);
    if (fs.existsSync(chapterFile)) return chapterFile;
    // 默认章 story.json 若即该章则兜底
    const defaultFile = path.join(STORIES_DIR, storyId, "story.json");
    if (fs.existsSync(defaultFile)) {
      try {
        const p = JSON.parse(fs.readFileSync(defaultFile, "utf8")) as { chapter_id?: string };
        if (p?.chapter_id === cid) return defaultFile;
      } catch {
        /* fallthrough */
      }
    }
    throw new StoryError(`章节不存在: ${storyId}/${cid}（可在「章节管理」新建）`);
  }
  return path.join(STORIES_DIR, storyId, "story.json");
}

/** 新章节的目标文件（chapters/<chapter_id>.json），创建/保存时用。 */
export function chapterFilePath(storyId: string, chapterId: string): string {
  return path.join(STORIES_DIR, storyId, "chapters", `${chapterId}.json`);
}

export function loadStory(storyId: string, chapterId?: string): { pack: StoryPack; path: string } {
  if (!storyId?.trim()) {
    throw new StoryError("必须显式指定 story_id（一故事一剧本，禁止无引用成片）");
  }
  const p = storyPath(storyId, chapterId);
  if (!fs.existsSync(p)) {
    throw new StoryError(`故事不存在: ${p}`);
  }
  const pack = JSON.parse(fs.readFileSync(p, "utf8")) as StoryPack;
  if (pack.story_id !== storyId) {
    throw new StoryError(
      `故事引用不匹配：请求 ${storyId}，文件 story_id=${pack.story_id}`,
    );
  }
  if (chapterId?.trim() && pack.chapter_id !== chapterId.trim()) {
    throw new StoryError(
      `章节引用不匹配：请求 ${chapterId}，文件 chapter_id=${pack.chapter_id}`,
    );
  }
  pack.props = pack.props || [];
  const issues = validateStory(pack);
  if (issues.length) {
    throw new StoryError("story pack invalid:\n- " + issues.join("\n- "));
  }
  return { pack, path: p };
}

export type ChapterSummary = {
  chapter_id: string;
  title?: string;
  logline?: string;
  project_id?: string;
  path: string;
  is_default: boolean;
  draft: boolean;
  shots: number;
  episodes: number;
};

/** 列出故事全部章节：默认章 story.json + chapters/*.json。 */
export function listChapters(storyId: string): ChapterSummary[] {
  const dir = path.join(STORIES_DIR, storyId);
  if (!fs.existsSync(dir)) {
    throw new StoryError(`故事不存在: ${dir}`);
  }
  const out: ChapterSummary[] = [];
  const readChapter = (file: string, isDefault: boolean) => {
    try {
      const p = JSON.parse(fs.readFileSync(file, "utf8")) as StoryPack;
      if (p?.story_id !== storyId) return;
      out.push({
        chapter_id: p.chapter_id || (isDefault ? "CH01" : path.basename(file, ".json")),
        title: p.title,
        logline: (p.script as { logline?: string })?.logline,
        project_id: p.project_id,
        path: toRepoRelative(file),
        is_default: isDefault,
        draft: p.draft === true,
        shots: Array.isArray(p.shots) ? p.shots.length : 0,
        episodes: Array.isArray((p.script as { episodes?: unknown[] })?.episodes)
          ? ((p.script as { episodes?: unknown[] }).episodes as unknown[]).length
          : 0,
      });
    } catch {
      /* skip unreadable file */
    }
  };
  const defaultFile = path.join(dir, "story.json");
  if (fs.existsSync(defaultFile)) readChapter(defaultFile, true);
  const chaptersDir = path.join(dir, "chapters");
  if (fs.existsSync(chaptersDir)) {
    for (const f of fs
      .readdirSync(chaptersDir)
      .filter((f) => f.endsWith(".json"))
      .sort()) {
      readChapter(path.join(chaptersDir, f), false);
    }
  }
  if (!out.length) {
    throw new StoryError(`故事无可用章节: ${storyId}`);
  }
  return out.sort((a, b) => a.chapter_id.localeCompare(b.chapter_id));
}

export type CreateChapterOptions = {
  /** 从哪一章克隆宇宙（人物/环境/道具/画风）；缺省为默认章 */
  from_chapter?: string;
  title?: string;
  logline?: string;
  synopsis?: string;
  project_id?: string;
};

/** 新建章节：克隆宇宙资产，剧本/分镜独立（draft，待后续流程填充）。 */
export function createChapter(
  storyId: string,
  chapterId: string,
  opts: CreateChapterOptions = {},
): { pack: StoryPack; path: string } {
  const cid = chapterId.trim().toUpperCase();
  if (!/^CH\d{2,}$/.test(cid)) {
    throw new StoryError(`章节 ID 需形如 CH02：${chapterId}`);
  }
  const target = chapterFilePath(storyId, cid);
  if (fs.existsSync(target)) {
    throw new StoryError(`章节已存在: ${storyId}/${cid}`);
  }
  const { pack: src } = loadStory(storyId, opts.from_chapter);
  const pack: StoryPack = {
    story_id: storyId,
    chapter_id: cid,
    project_id: opts.project_id || `${src.project_id || storyId}_${cid.toLowerCase()}`,
    title: opts.title || `${src.title || storyId}·${cid}`,
    style_lock: src.style_lock,
    resolution: src.resolution,
    video_workflow: src.video_workflow,
    steps: src.steps,
    draft: true,
    script: {
      logline: opts.logline || "",
      synopsis: opts.synopsis || "",
      episodes: [],
    },
    characters: structuredClone(src.characters),
    environments: structuredClone(src.environments),
    props: structuredClone(src.props || []),
    shots: [],
  };
  saveStory(pack, target);
  ok("章节管理", "已创建章节", { story_id: storyId, chapter_id: cid, path: target });
  return { pack, path: target };
}

export function validateStory(pack: StoryPack): string[] {
  const issues: string[] = [];
  if (!pack.story_id) issues.push("missing story_id");
  if (!pack.chapter_id) issues.push("missing chapter_id");
  if (!pack.script) issues.push("missing script");
  if (!pack.characters?.length) issues.push("characters required");
  if (!pack.environments?.length) issues.push("environments required");
  // draft 章节（章节管理新建、待填充）允许空分镜
  if (!pack.draft && !pack.shots?.length) issues.push("shots required");

  const charIds = new Set(pack.characters.map((c) => c.id));
  const envIds = new Set(pack.environments.map((e) => e.id));
  const propIds = new Set((pack.props || []).map((p) => p.id));
  const shotIds = new Set(pack.shots.map((s) => s.shot_id));

  for (const sh of pack.shots) {
    if (!envIds.has(sh.environment_id)) {
      issues.push(`${sh.shot_id}: unknown environment_id ${sh.environment_id}`);
    }
    for (const cid of sh.character_ids || []) {
      if (!charIds.has(cid)) issues.push(`${sh.shot_id}: unknown character_id ${cid}`);
    }
    for (const pid of sh.prop_ids || []) {
      if (!propIds.has(pid)) issues.push(`${sh.shot_id}: unknown prop_id ${pid}`);
    }
    for (const url of sh.ref_images || []) {
      if (url === "") issues.push(`${sh.shot_id}: empty string in ref_images (forbidden)`);
    }
    if (sh.bridge_from && !shotIds.has(String(sh.bridge_from))) {
      issues.push(`${sh.shot_id}: bridge_from unknown ${sh.bridge_from}`);
    }
  }
  return issues;
}

export function saveStory(pack: StoryPack, filePath?: string): string {
  const p = filePath || storyPath(pack.story_id);
  ensureDir(path.dirname(p));
  pack.props = pack.props || [];
  fs.writeFileSync(p, JSON.stringify(pack, null, 2), "utf8");
  registerStory(pack, p);
  return p;
}

function toRepoRelative(filePath: string): string {
  const abs = path.resolve(filePath);
  const root = path.resolve(REPO_ROOT);
  const rel = path.relative(root, abs).replace(/\\/g, "/");
  if (rel.startsWith("..") || path.isAbsolute(rel)) {
    warn("故事注册", "路径不在仓库内，仍写入原路径", { path: abs });
    return abs.replace(/\\/g, "/");
  }
  return rel;
}

function registerStory(pack: StoryPack, filePath: string): void {
  const cfg = loadConfig();
  const indexPath = path.join(cfg.storiesDir, "index.json");
  ensureDir(cfg.storiesDir);
  let idx: { stories: Record<string, unknown> } = { stories: {} };
  if (fs.existsSync(indexPath)) {
    idx = JSON.parse(fs.readFileSync(indexPath, "utf8"));
    idx.stories ||= {};
  }
  const rel = toRepoRelative(filePath);
  const logline = (pack.script as { logline?: string })?.logline;
  const entry = ((idx.stories as Record<string, Record<string, unknown>>)[pack.story_id] ||
    {}) as Record<string, unknown>;
  // 故事→章节两级索引；顶层字段保留为"最近注册章"以兼容旧消费方
  const chapters = (entry.chapters || {}) as Record<string, unknown>;
  if (pack.chapter_id) {
    chapters[pack.chapter_id] = { project_id: pack.project_id, path: rel, logline };
  }
  entry.chapters = chapters;
  entry.title = pack.title || (entry.title as string) || pack.story_id;
  entry.chapter_id = pack.chapter_id;
  entry.project_id = pack.project_id;
  entry.path = rel;
  entry.logline = logline;
  (idx.stories as Record<string, unknown>)[pack.story_id] = entry;
  fs.writeFileSync(indexPath, JSON.stringify(idx, null, 2), "utf8");
  ok("故事注册", "已写入索引", { story_id: pack.story_id, chapter_id: pack.chapter_id, path: rel });
}

/** 宇宙同步：把源章的人物/环境/道具/画风锁复制到故事其余各章。 */
export function syncUniverse(storyId: string, fromChapter?: string): { updated: string[] } {
  const { pack: src } = loadStory(storyId, fromChapter);
  const updated: string[] = [];
  for (const ch of listChapters(storyId)) {
    if (ch.chapter_id === src.chapter_id) continue;
    const { pack, path: p } = loadStory(storyId, ch.chapter_id);
    pack.characters = structuredClone(src.characters);
    pack.environments = structuredClone(src.environments);
    pack.props = structuredClone(src.props || []);
    pack.style_lock = src.style_lock;
    saveStory(pack, p);
    updated.push(ch.chapter_id);
  }
  if (updated.length) {
    ok("章节管理", "宇宙已同步", { story_id: storyId, from: src.chapter_id, to: updated.join(",") });
  }
  return { updated };
}

export function getCharacter(pack: StoryPack, characterId: string) {
  const ch = pack.characters.find((c) => c.id === characterId);
  if (!ch) throw new StoryError(`unknown character_id: ${characterId}`);
  return ch;
}

export function getShot(pack: StoryPack, shotId: string) {
  const sh = pack.shots.find((s) => s.shot_id === shotId);
  if (!sh) throw new StoryError(`unknown shot_id: ${shotId}`);
  return sh;
}

export function getEnvironment(pack: StoryPack, envId: string) {
  const env = pack.environments.find((e) => e.id === envId);
  if (!env) throw new StoryError(`unknown environment_id: ${envId}`);
  return env;
}

export function getProp(pack: StoryPack, propId: string) {
  const prop = (pack.props || []).find((p) => p.id === propId);
  if (!prop) throw new StoryError(`unknown prop_id: ${propId}`);
  return prop;
}

export function loadStyleLock(name?: string): string {
  if (!name) return "";
  const p = path.join(loadConfig().repoRoot, "templates", "style_locks", `${name}.txt`);
  if (fs.existsSync(p)) return fs.readFileSync(p, "utf8").trim();
  return name;
}

export function listStyleLocks(): string[] {
  const dir = path.join(loadConfig().repoRoot, "templates", "style_locks");
  if (!fs.existsSync(dir)) return [];
  return fs
    .readdirSync(dir)
    .filter((f) => f.endsWith(".txt"))
    .map((f) => f.replace(/\.txt$/, ""));
}

export function withLocks(prompt: string, styleLock: string, identityLock: string): string {
  return [styleLock, identityLock, prompt].filter(Boolean).join("。");
}

File: src/story.py
# -*- coding: utf-8 -*-
"""StoryPack: 剧本 → 人物 → 环境 → 分镜 → ShotJob。"""
from __future__ import annotations

from typing import Any

try:
    from constants import DEFAULT_RESOLUTION, DEFAULT_STYLE_LOCK, DEFAULT_WORKFLOW
except ImportError:  # 包方式导入（import src.story）
    from .constants import DEFAULT_RESOLUTION, DEFAULT_STYLE_LOCK, DEFAULT_WORKFLOW


class StoryError(ValueError):
    pass


def _nonempty_urls(*values: str | None) -> list[str]:
    out: list[str] = []
    seen: set[str] = set()
    for v in values:
        if not v or not str(v).strip():
            continue
        u = str(v).strip()
        if u in seen:
            continue
        seen.add(u)
        out.append(u)
    return out


def index_by_id(items: list[dict[str, Any]], key: str = "id") -> dict[str, dict[str, Any]]:
    out: dict[str, dict[str, Any]] = {}
    for it in items:
        iid = it.get(key)
        if not iid:
            raise StoryError(f"missing {key} in item: {it}")
        if iid in out:
            raise StoryError(f"duplicate {key}: {iid}")
        out[str(iid)] = it
    return out


def validate_story_pack(pack: dict[str, Any]) -> list[str]:
    """Return human-readable issues; empty list means OK."""
    issues: list[str] = []
    if not pack.get("story_id"):
        issues.append("missing story_id（一故事一剧本，成片必须引用）")
    if not pack.get("project_id"):
        issues.append("missing project_id")
    if not pack.get("chapter_id"):
        issues.append("missing chapter_id（如 CH01）")
    script = pack.get("script") or {}
    if not script.get("logline"):
        issues.append("script.logline required")
    if not script.get("synopsis"):
        issues.append("script.synopsis required")
    if not script.get("episodes"):
        issues.append("script.episodes required")

    chars = index_by_id(list(pack.get("characters") or []))
    envs = index_by_id(list(pack.get("environments") or []))
    props = index_by_id(list(pack.get("props") or []))
    shots = list(pack.get("shots") or [])
    if not chars:
        issues.append("characters required")
    if not envs:
        issues.append("environments required")
    if not shots:
        issues.append("shots required")

    ep_ids = {e.get("episode_id") for e in (script.get("episodes") or [])}
    shot_ids: set[str] = set()
    for sh in shots:
        sid = sh.get("shot_id")
        if not sid:
            issues.append("shot missing shot_id")
            continue
        if sid in shot_ids:
            issues.append(f"duplicate shot_id: {sid}")
        shot_ids.add(sid)
        if sh.get("episode_id") not in ep_ids:
            issues.append(f"{sid}: episode_id not in script.episodes")
        if sh.get("environment_id") not in envs:
            issues.append(f"{sid}: unknown environment_id {sh.get('environment_id')}")
        for cid in sh.get("character_ids") or []:
            if cid not in chars:
                issues.append(f"{sid}: unknown character_id {cid}")
        for pid in sh.get("prop_ids") or []:
            if pid not in props:
                issues.append(f"{sid}: unknown prop_id {pid}")
        if not (sh.get("action") or sh.get("video_prompt")):
            issues.append(f"{sid}: need action or video_prompt")
        plan = sh.get("plan_path")
        if plan and plan not in ("video_ref", "bridge", "grid", "lipsync"):
            issues.append(f"{sid}: invalid plan_path {plan}")
        if sh.get("grid_cell") and (plan or "video_ref") != "grid":
            issues.append(f"{sid}: grid_cell 仅当 plan_path=grid 时使用")
        for url in sh.get("ref_images") or []:
            if url == "":
                issues.append(f"{sid}: empty string in ref_images (forbidden)")

    for sh in shots:
        bf = sh.get("bridge_from")
        if bf and bf not in shot_ids:
            issues.append(f"{sh.get('shot_id')}: bridge_from unknown {bf}")

    return issues


def character_identity_lock(chars: dict[str, dict[str, Any]], character_ids: list[str]) -> str:
    parts: list[str] = []
    for cid in character_ids:
        ch = chars.get(cid)
        if not ch:
            raise StoryError(f"unknown character_id: {cid}")
        lock = (ch.get("identity_lock") or "").strip()
        if not lock:
            raise StoryError(f"character {cid} missing identity_lock")
        parts.append(lock)
    return "；".join(parts)


def assemble_ref_images(
    pack: dict[str, Any],
    shot: dict[str, Any],
    chars: dict[str, dict[str, Any]],
    envs: dict[str, dict[str, Any]],
) -> list[str]:
    if shot.get("ref_images"):
        return _nonempty_urls(*list(shot["ref_images"]))

    urls: list[str] = []
    for cid in shot.get("character_ids") or []:
        refs = (chars[cid].get("ref_images") or {}) if cid in chars else {}
        urls.extend(
            _nonempty_urls(
                refs.get("sheet"),
                refs.get("face"),
                refs.get("full"),
                refs.get("costume"),
            )
        )
    props = index_by_id(list(pack.get("props") or []))
    for pid in shot.get("prop_ids") or []:
        prefs = (props[pid].get("ref_images") or {}) if pid in props else {}
        urls.extend(_nonempty_urls(prefs.get("sheet"), prefs.get("detail")))
    still = shot.get("still_url")
    if still:
        urls.extend(_nonempty_urls(still))
    env = envs.get(shot.get("environment_id") or "")
    if env:
        erefs = env.get("ref_images") or {}
        urls.extend(_nonempty_urls(erefs.get("establishing"), erefs.get("detail")))
    # also allow pack-level shared refs
    urls.extend(_nonempty_urls(*(pack.get("ref_images") or [])))
    return _nonempty_urls(*urls)


def compile_camera_action(shot: dict[str, Any]) -> str:
    if shot.get("video_prompt"):
        return str(shot["video_prompt"]).strip()
    duration = int(shot.get("duration") or 5)
    cam = shot.get("camera") or {}
    bits = [
        f"0-{duration}秒",
        shot.get("action") or "",
    ]
    size = cam.get("shot_size")
    move = cam.get("move")
    angle = cam.get("angle")
    cam_bits = [x for x in [size and f"景别{size}", move and f"运镜{move}", angle and f"机位{angle}"] if x]
    if cam_bits:
        bits.append("，".join(cam_bits))
    if shot.get("emotion"):
        bits.append(f"情绪：{shot['emotion']}")
    if shot.get("dialogue"):
        bits.append(f"对白：「{shot['dialogue']}」")
    return "：".join([bits[0], "，".join(b for b in bits[1:] if b)])


def compile_still_prompt(
    shot: dict[str, Any],
    env: dict[str, Any],
    identity_lock: str,
) -> str:
    if shot.get("still_prompt"):
        return str(shot["still_prompt"]).strip()
    cam = shot.get("camera") or {}
    parts = [
        env.get("scene_card") or "",
        shot.get("action") or "",
        identity_lock,
    ]
    if cam.get("shot_size"):
        parts.insert(0, f"{cam['shot_size']}景")
    if shot.get("emotion"):
        parts.append(f"情绪：{shot['emotion']}")
    return "。".join(p for p in parts if p)


def expand_shot(
    pack: dict[str, Any],
    shot: dict[str, Any],
    *,
    chars: dict[str, dict[str, Any]] | None = None,
    envs: dict[str, dict[str, Any]] | None = None,
) -> dict[str, Any]:
    """Expand one story shot into a runnable video ShotJob (+ metadata)."""
    chars = chars or index_by_id(list(pack.get("characters") or []))
    envs = envs or index_by_id(list(pack.get("environments") or []))
    env_id = shot.get("environment_id")
    if env_id not in envs:
        raise StoryError(f"unknown environment_id: {env_id}")
    env = envs[env_id]
    cids = list(shot.get("character_ids") or [])
    identity = character_identity_lock(chars, cids)
    refs = assemble_ref_images(pack, shot, chars, envs)
    duration = int(shot.get("duration") or 5)
    video_prompt = compile_camera_action(shot)
    still_prompt = compile_still_prompt(shot, env, identity)

    job: dict[str, Any] = {
        "shot_id": shot["shot_id"],
        "id": shot["shot_id"],
        "episode_id": shot.get("episode_id"),
        "scene_id": shot.get("scene_id"),
        "beat_id": shot.get("beat_id"),
        "environment_id": env_id,
        "character_ids": cids,
        "prop_ids": list(shot.get("prop_ids") or []),
        "duration": duration,
        "resolution": pack.get("resolution") or DEFAULT_RESOLUTION,
        "workflow": shot.get("workflow") or pack.get("video_workflow") or DEFAULT_WORKFLOW,
        "plan_path": shot.get("plan_path") or "video_ref",
        "prompt": video_prompt,
        "still_prompt": still_prompt,
        "identity_lock": identity,
        "ref_images": refs,
        "tail_frame_note": shot.get("tail_frame_note"),
        "state": dict(shot.get("state") or {}),
        "bridge_from": shot.get("bridge_from"),
        "needs_lipsync": bool(shot.get("needs_lipsync")),
        "dialogue": shot.get("dialogue"),
    }
    if shot.get("seed") is not None:
        job["seed"] = shot["seed"]
    if shot.get("ref_audios"):
        job["ref_audios"] = list(shot["ref_audios"])
    if shot.get("first_frame"):
        job["first_frame"] = shot["first_frame"]
    if shot.get("last_frame"):
        job["last_frame"] = shot["last_frame"]
    if shot.get("still_url"):
        job["still_url"] = shot["still_url"]
    if shot.get("audio_duration") is not None:
        job["audio_duration"] = shot["audio_duration"]
    for cid in cids:
        vr = (chars.get(cid) or {}).get("voice_ref")
        if vr:
            job["voice_ref"] = vr
            break
    # continuity defaults from environment
    st = job["state"]
    st.setdefault("location", env.get("name"))
    if env.get("time_of_day"):
        st.setdefault("time_of_day", env["time_of_day"])
    if env.get("weather"):
        st.setdefault("weather", env["weather"])
    return job


def expand_story_pack(pack: dict[str, Any]) -> dict[str, Any]:
    issues = validate_story_pack(pack)
    if issues:
        raise StoryError("story pack invalid:\n- " + "\n- ".join(issues))
    chars = index_by_id(list(pack["characters"]))
    envs = index_by_id(list(pack["environments"]))
    jobs: list[dict[str, Any]] = []
    prev_state: dict[str, Any] | None = None
    for sh in pack["shots"]:
        job = expand_shot(pack, sh, chars=chars, envs=envs)
        if prev_state:
            job["state_prev"] = prev_state
        prev_state = dict(job.get("state") or {})
        jobs.append(job)
    return {
        "story_id": pack["story_id"],
        "chapter_id": pack.get("chapter_id"),
        "project_id": pack["project_id"],
        "title": pack.get("title"),
        "style_lock": pack.get("style_lock") or DEFAULT_STYLE_LOCK,
        "resolution": pack.get("resolution") or DEFAULT_RESOLUTION,
        "video_workflow": pack.get("video_workflow") or DEFAULT_WORKFLOW,
        "steps": pack.get("steps") or ["video"],
        "script": pack["script"],
        "shot_jobs": jobs,
    }


def derive_bridge_frames(
    job: dict[str, Any],
    jobs_by_id: dict[str, dict[str, Any]],
) -> tuple[str | None, str | None]:
    """Bridge 首尾帧派生：显式帧优先，缺失时回退到相邻镜静帧。

    - first_frame ← 本镜显式 first_frame，否则上一镜（bridge_from）的 last_frame/still_url
    - last_frame  ← 本镜显式 last_frame，否则本镜 still_url
    """
    src = jobs_by_id.get(str(job.get("bridge_from") or "")) or {}
    first = job.get("first_frame") or src.get("last_frame") or src.get("still_url")
    last = job.get("last_frame") or job.get("still_url")
    return (first or None), (last or None)


def assert_production_approved(pack: dict[str, Any], shot_ids: list[str] | None = None) -> None:
    """成片审批门闸：未批准的角色定妆 / 静帧禁止进入成片（对齐 dsh gate.ts）。"""
    chars = index_by_id(list(pack.get("characters") or []))
    allow = set(shot_ids) if shot_ids else None
    problems: list[str] = []
    for sh in pack.get("shots") or []:
        sid = sh.get("shot_id")
        if allow is not None and sid not in allow:
            continue
        for cid in sh.get("character_ids") or []:
            ch = chars.get(cid)
            if ch and ch.get("approved") is False:
                problems.append(
                    f"{sid}: 角色 {cid}（{ch.get('name') or cid}）定妆未批准 approved=false"
                )
        if sh.get("still_url") and sh.get("still_approved") is False:
            problems.append(f"{sid}: 静帧未批准 still_approved=false")
    if problems:
        raise StoryError("审批门闸未通过（请先在工作台批准）：\n- " + "\n- ".join(problems))


def validate_state_continuity(pack: dict[str, Any]) -> list[str]:
    """相邻镜 state 连续性检查：返回警告列表（不阻断，供复盘）。"""
    warnings: list[str] = []
    prev: tuple[str, dict[str, Any]] | None = None
    for sh in pack.get("shots") or []:
        sid = str(sh.get("shot_id") or "?")
        st = dict(sh.get("state") or {})
        if prev is not None:
            psid, pst = prev
            for key in ("wardrobe", "holding", "facing"):
                pv, cv = pst.get(key), st.get(key)
                if pv and cv and pv != cv:
                    warnings.append(f"{sid}: state.{key} 与前镜 {psid} 不连续（{pv} → {cv}）")
        prev = (sid, st)
    return warnings
